torch/csrc/autograd/VariableTypeManual.cpp

```diff
@@ -345,4 +345,75 @@ Tensor & VariableType::detach_(Tensor & self) const {
   return self;
 }
 
+Tensor VariableType::_sparse_coo_tensor_with_dims_and_tensors(int64_t sparse_dim, int64_t dense_dim, IntArrayRef size, const Tensor & indices, const Tensor & values, const TensorOptions & options) const {
+  RECORD_FUNCTION("_sparse_coo_tensor_with_dims_and_tensors", std::vector<c10::IValue>({indices, values}), Function::peek_at_next_sequence_nr());
+
+  // SparseTensorImpl's `set_indices_and_values_unsafe(indices, values)` function expects
+  // `indices` and `values` to not require grad. The `unpack()` function calls in the
+  // auto-generated version of this function doesn't always return `indices_` and `values_`
+  // with requires_grad=false, and we need another way to create non-requires-grad `indices_`
+  // and `values_` tensors that contains the same data as `indices` and `values`, which is
+  // achieved by using `indices`'s and `values`'s `tensor_data()`.
+  auto indices_ = as_variable_ref(indices).tensor_data();
+  auto values_ = as_variable_ref(values).tensor_data();
```

Comment:
The only change in the implementation of `VariableType::_sparse_coo_tensor_with_dims_and_tensors` is these two lines for `indices_` and `values_` (they used to be `auto& indices_ = unpack(indices, "indices", 3)` and `auto& values_ = unpack(values, "values", 4)`). We need this manual implementation because of the following reason:

SparseTensorImpl's `set_indices_and_values_unsafe(indices, values)` function expects `indices` and `values` to not requires grad. Before this patch, we use `unpack()` to unwrap the Variable and obtain the underlying data tensor, which is guaranteed to not requires grad. After this patch, `unpack()` doesn't do the unwrapping anymore, and we need a way to create non-requires-grad `indices_` and `values_` tensors that contains the same data as `indices` and `values`, which is achieved by using `indices`'s and `values`'s `tensor_data()`.

Reply:
is the problem that _sparse_coo_tensor_with_dims_and_tensors checks `!is_variable`?  This is just a consequence of making the optimization that unpack doesn't call `tensor_data`, right?

Reply:
one thing to consider here is how you want to solve this once there is no real distinction between variables and tensors.

Putting this in VariableTypeManual seems wrong.  We can't put magic for every function that wants the non-autograd version of a tensor.

Reply:
> is the problem that _sparse_coo_tensor_with_dims_and_tensors checks `!is_variable`? This is just a consequence of making the optimization that unpack doesn't call `tensor_data`, right?

Yes that's the exact reason.

> one thing to consider here is how you want to solve this once there is no real distinction between variables and tensors.

> Putting this in VariableTypeManual seems wrong. We can't put magic for every function that wants the non-autograd version of a tensor.

We can put this function in a special list in `gen_variable_type.py`, and the codegen can look at this list to know the functions that want the non-autograd version of a tensor.

Reply:
Another aspect we might need to discuss is whether we really should enforce the invariant that `values_` in a sparse tensor does not require grad. I think since the public API `values()` always ignores `values_`'s original AutogradMeta and always creates a new AutogradMeta for the differentiable view:
```cpp
Tensor VariableType::values(const Tensor & self) const {
  ...
  auto tmp = ([&]() {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    return baseType->values(self_);
  })();
  auto result = as_view(self, tmp, true);
  ...
}
```
```cpp
inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable = true) {
  auto base_var = Variable(base);
  if (base_var.is_view()) {
    base_var = base_var.base();
  }
  return make_variable_view(std::move(base_var), std::move(tensor), is_differentiable);
}
```
```cpp
inline Variable make_variable_view(
    Variable base,
    at::Tensor data,
    bool is_differentiable = true,
    bool allow_tensor_metadata_change = true,
    Edge gradient_edge = Edge()) {
  if (data.defined()) {
    if (is_differentiable) {
      /// Differentiable view. Track history with DifferentiableViewMeta.
      auto data_impl_copy = data.getIntrusivePtr()->shallow_copy_and_detach();
      data_impl_copy->set_allow_tensor_metadata_change(allow_tensor_metadata_change);
      data_impl_copy->set_autograd_meta(c10::guts::make_unique<Variable::DifferentiableViewMeta>(
        data_impl_copy.get(), std::move(base), std::move(gradient_edge)));
      return Variable(data_impl_copy);
    } else {
      ...
    }
  }
  return Variable();
}
```
we don't need to care about whether a sparse tensor's `values_` contains AutogradMeta.

Reply:
Simon asked in https://github.com/pytorch/pytorch/pull/20295#issuecomment-490738284:
> So a `values_` tensor will have the `AutogradMetaData`, but it is just ignored, for the computations the sparse tensor involves. That's great. But I wonder what happens if users later, say, `resize_` or `as_strided_` on the same value tensor that is also now kept in the SparseTensorImpl?

Calling `resize_ `or `as_strided_` on the same value tensor is likely not a supported operation in current master. Here are a few examples:
```python
# Calling resize_ on requires-grad value tensor
i1 = torch.zeros([1, 1])
v1 = torch.ones([1, 2, 3]).requires_grad_()
t1 = torch.sparse_coo_tensor(i1, v1, torch.Size([2, 2, 3]))
v1.resize_(4, 5) # throws "cannot resize variables that require grad"

# Calling resize_ on non-requires-grad value tensor
i2 = torch.zeros([1, 1])
v2 = torch.ones([1, 2, 3])
t2 = torch.sparse_coo_tensor(i2, v2, torch.Size([2, 2, 3]))
v2.resize_(4, 5) # breaks sparse shape invariant
t2.coalesce().values() # throws "indices and values must have same nnz, but got nnz from indices: 1, nnz from values: 4"

# Calling as_strided_ on requires-grad value tensor
i3 = torch.zeros([1, 1])
v3 = torch.ones([1, 2, 3]).requires_grad_()
t3 = torch.sparse_coo_tensor(i3, v3, torch.Size([2, 2, 3]))
v3.as_strided_((2, 3), (3, 1)) # throws "a leaf Variable that requires grad has been used in an in-place operation"

# Calling as_strided_ on non-requires-grad value tensor
i4 = torch.zeros([1, 1])
v4 = torch.ones([1, 2, 3])
t4 = torch.sparse_coo_tensor(i4, v4, torch.Size([2, 2, 3]))
v4.as_strided_((2, 3), (3, 1)) # breaks sparse shape invariant
t4.coalesce().values() # throws "indices and values must have same nnz, but got nnz from indices: 1, nnz from values: 2"
```

cc. @SsnL